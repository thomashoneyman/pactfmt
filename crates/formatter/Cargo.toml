[package]
name = "formatter"
version = "0.1.0"
edition = "2021"

[dependencies]
syntax = { path = "../syntax" }
pretty = "0.12.3"

[dev-dependencies]
insta = "1.41.1"
pretty_assertions = "1.4.0"
test_each_file = "0.3.4"
test-utils = { path = "../test-utils" }
