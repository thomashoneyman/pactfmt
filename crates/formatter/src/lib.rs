mod format_doc;
mod format_tree;
mod lower_tree;

use ariadne::{Label, Report, ReportKind, Source};
use line_index::{LineCol, LineIndex};
use pretty::RcAllocator;
use syntax::parser::{parse, ParseError};
use syntax::{report_errors, tokenize};

use crate::{format_doc::FormatDoc, format_tree::FST};

/// Format the input source code, returning an error if the parsed input
/// contains any error trees.
pub fn format_source(input: &str, width: usize) -> Result<String, Vec<u8>> {
    let tokens = tokenize(input);
    let (parsed_trees, parsed_errors) = parse(tokens);
    let significant_errors = report_errors(&parsed_trees, &parsed_errors);

    if !significant_errors.is_empty() {
        return Err(format_errors(input, significant_errors));
    }

    let fst_trees: Vec<FST> = parsed_trees
        .into_iter()
        .map(lower_tree::lower_tree)
        .collect();

    let allocator = RcAllocator;
    let formatted = fst_trees
        .iter()
        .fold(FormatDoc::nil(&allocator), |acc, fst| {
            acc.append(fst.format(&allocator))
        })
        .pretty(width);

    Ok(formatted)
}

fn format_errors(input: &str, errors: Vec<&ParseError>) -> Vec<u8> {
    let line_index = LineIndex::new(input);
    let errors = errors.iter().filter_map(|&error| {
        let line = (error.line - 1) as u32;
        let col = (error.column - 1) as u32;
        let line_col = LineCol { line, col };
        let index: usize = line_index.offset(line_col)?.into();
        Some((&error.message, index))
    });

    let mut report =
        Report::build(ReportKind::Error, ("<stdin>", 0..input.len())).with_message("Syntax error!");

    for (message, index) in errors {
        report = report.with_label(Label::new(("<stdin>", index..index)).with_message(message))
    }

    let mut buffer = vec![];
    report
        .finish()
        .write(("<stdin>", Source::from(input)), &mut buffer)
        .unwrap();
    buffer
}

/// True if the input can be formatted (no parse errors) and the formatted
/// output is the same as the input.
pub fn check_source(input: &str, width: usize) -> bool {
    let formatted = format_source(input, width);
    match formatted {
        Ok(formatted) => input == formatted,
        Err(_) => false,
    }
}
