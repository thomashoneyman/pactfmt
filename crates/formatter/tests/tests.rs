use formatter::format_source;
use test_each_file::test_each_file;

test_each_file! { in "crates/formatter/tests/fixtures" => |content: &str| {
    insta::assert_snapshot!(format_source(content, 40));
}}
