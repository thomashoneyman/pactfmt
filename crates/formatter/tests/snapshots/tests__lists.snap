---
source: crates/formatter/tests/tests.rs
expression: "format_source(content, 40)"
---
[1 2 3]
[1 2 3]
[1 2 3]
[
  1 ;hm
  ;hm
  2
]
